<subsystem_configuration>
    <import package="controller_common" />
    <import package="rtt_geometry_msgs" />
    <import package="rtt_velma_core_cs_ve_body_msgs" />
    <import package="rtt_velma_core_cs_ve_body_subsystem_ports" />
    <import package="velma_core_ve_body" />
    <import package="eigen_typekit" />
    <import package="rtt_std_msgs" />
    <import package="port_operations" />
    <import package="rtt_control_msgs" />
    <import package="velma_controller" />
    <import package="rtt_cartesian_trajectory_msgs" />
    <import package="rtt_std_msgs" />
    <import package="rtt_tf" />
    <import package="rtt_test_ec_subsystem_ports" />
    <import package="rtt_test_ec_msgs" />
    <import package="rtt_lwr_msgs" />
    <import package="barrett_hand_controller" />
    <import package="ec_can_queue" />

    <component name="TorsoPanTransmision"       type="SimpleTransmision" running="true" />
    <component name="TorsoPanTransmisionInv"    type="SimpleTransmisionInv" />
    <component name="HeadPanTransmision"        type="SimpleTransmision" running="true" />
    <component name="HeadPanTransmisionInv"     type="SimpleTransmisionInv" />
    <component name="HeadTiltTransmision"       type="SimpleTransmision" running="true" />
    <component name="HeadTiltTransmisionInv"    type="SimpleTransmisionInv" />

    <component name="bypass" type="velma_core_ve_body_types::BypassComponent" />
    <component name="safe" type="velma_core_ve_body_types::SafeComponent" />
    <component name="packet_counter" type="velma_core_ve_body_types::PacketCounterComponent" running="true" />
    <component name="hw_state" type="velma_core_ve_body_types::HwState" running="true" />

    <component name="rHand" type="BarrettHand" running="true" >
        <service name="can_queue" />
    </component>

    <component name="lHand" type="BarrettHand" running="true" >
        <service name="can_queue" />
    </component>

    <component name="can_queue_tx" type="ec_can_queue::CanQueueTxComponent" running="true" />

    <component name="ZeroVec8"  type="VectorDummy8" latex="\text{ZeroVec}_{8}" running="true" />
    <component name="Zero"  type="ScalarZeroUInt32" latex="\text{Zero}" running="true" />

    <io_buffer alias="hi_st" name="velma_core_cs_ve_body_msgs_Status" />
    <io_buffer alias="hi_cmd" name="velma_core_cs_ve_body_msgs_Command" />

    <io_buffer alias="lo_rLwr_st" name="rLwr_st" />
    <io_buffer alias="lo_rLwr_cmd" name="rLwr_cmd" />

    <io_buffer alias="lo_lLwr_st" name="lLwr_st" />
    <io_buffer alias="lo_lLwr_cmd" name="lLwr_cmd" />

    <io_buffer alias="lo_ec_st" name="EC_Status" />
    <io_buffer alias="lo_ec_cmd" name="EC_Command" />

    <ros_stream port="master_component.hi_cmd_OUTPORT"      topic="/velma_core_ve_body/hi_cmd" />

    <ros_stream port="hi_stConcate.msg_OUTPORT"             topic="/velma_core_ve_body/hi_st" />

    <ros_stream port="master_component.lo_rLwr_st_OUTPORT"              topic="/velma_core_ve_body/lo_rLwr_st" />
    <ros_stream port="master_component.lo_lLwr_st_OUTPORT"              topic="/velma_core_ve_body/lo_lLwr_st" />
    <ros_stream port="master_component.lo_ec_st_OUTPORT"                topic="/velma_core_ve_body/lo_ec_st" />
    <ros_stream port="lo_ec_stRx.msg_OUTPORT"                           topic="/velma_core_ve_body/lo_ec_stRx" />

    <connection from="hi_cmdSplit.hpMotor_q_OUTPORT" to="HeadPanTransmisionInv.JointPosition_INPORT" />

    <connection from="hi_cmdSplit.htMotor_q_OUTPORT" to="HeadTiltTransmisionInv.JointPosition_INPORT" />

    <connection from="lo_ec_stSplit.HeadPan_Inputs_Positionactualvalue_OUTPORT"         to="HeadPanTransmision.MotorPosition_INPORT" />
    <connection from="lo_ec_stSplit.HeadPan_Inputs_Velocityactualvalue_OUTPORT"         to="HeadPanTransmision.MotorVelocity_INPORT" />
    <connection from="HeadPanTransmision.JointPosition_OUTPORT"         to="hi_stConcate.hpMotor_q_INPORT" />
    <connection from="HeadPanTransmision.JointVelocity_OUTPORT"         to="hi_stConcate.hpMotor_dq_INPORT" />

    <connection from="lo_ec_stSplit.HeadTilt_Inputs_Positionactualvalue_OUTPORT"        to="HeadTiltTransmision.MotorPosition_INPORT" />
    <connection from="lo_ec_stSplit.HeadTilt_Inputs_Velocityactualvalue_OUTPORT"        to="HeadTiltTransmision.MotorVelocity_INPORT" />
    <connection from="HeadTiltTransmision.JointPosition_OUTPORT" to="hi_stConcate.htMotor_q_INPORT" />
    <connection from="HeadTiltTransmision.JointVelocity_OUTPORT" to="hi_stConcate.htMotor_dq_INPORT" />

    <connection from="lo_ec_stSplit.TorsoPan_Inputs_Positionactualvalue_OUTPORT" to="TorsoPanTransmision.MotorPosition_INPORT" />
    <connection from="lo_ec_stSplit.TorsoPan_Inputs_Velocityactualvalue_OUTPORT" to="TorsoPanTransmision.MotorVelocity_INPORT" />

    <connection from="safe.sc_OUTPORT"              to="hi_stConcate.sc_INPORT" />
    <connection from="safe.rArm_t_OUTPORT"          to="lo_rLwr_cmdConcate.t_INPORT" />
    <connection from="safe.rArmFri_OUTPORT"         to="lo_rLwr_cmdConcate.krlCmd_INPORT" />
    <connection from="safe.lArm_t_OUTPORT"          to="lo_lLwr_cmdConcate.t_INPORT" />
    <connection from="safe.lArmFri_OUTPORT"         to="lo_lLwr_cmdConcate.krlCmd_INPORT" />
    <connection from="safe.tMotor_i_OUTPORT"        to="lo_ec_cmdConcate.TorsoPan_Outputs_TargetTorque_INPORT" />
    <connection from="safe.hpMotor_q_OUTPORT"       to="lo_ec_cmdConcate.HeadPan_Outputs_TargetPosition_INPORT" />
    <connection from="safe.htMotor_q_OUTPORT"       to="lo_ec_cmdConcate.HeadTilt_Outputs_TargetPosition_INPORT" />
    <connection from="lo_rLwr_stSplit.dq_OUTPORT"                    to="safe.rArm_dq_INPORT" name="dq" />
    <connection from="lo_rLwr_stSplit.iState_OUTPORT"                         to="safe.rArmFriIntf_INPORT" />
    <connection from="lo_rLwr_stSplit.rState_OUTPORT"                       to="safe.rArmFriRobot_INPORT" />
    <connection from="lo_lLwr_stSplit.dq_OUTPORT"                    to="safe.lArm_dq_INPORT" name="dq" />
    <connection from="lo_lLwr_stSplit.iState_OUTPORT"                         to="safe.lArmFriIntf_INPORT" />
    <connection from="lo_lLwr_stSplit.rState_OUTPORT"                       to="safe.lArmFriRobot_INPORT" />
    <connection from="lo_ec_stSplit.TorsoPan_Inputs_Velocityactualvalue_OUTPORT"                 to="safe.tMotor_dq_INPORT" name="dq_t" />
    <connection from="lo_ec_stSplit.HeadPan_Inputs_Positionactualvalue_OUTPORT"         to="safe.hpMotor_q_INPORT" name="q_hp" />
    <connection from="lo_ec_stSplit.HeadTilt_Inputs_Positionactualvalue_OUTPORT"        to="safe.htMotor_q_INPORT" name="q_ht" />

    <connection from="bypass.rArm_t_OUTPORT"        to="lo_rLwr_cmdConcate.t_INPORT" />
    <connection from="bypass.lArm_t_OUTPORT"        to="lo_lLwr_cmdConcate.t_INPORT" />
    <connection from="bypass.tMotor_i_OUTPORT"      to="lo_ec_cmdConcate.TorsoPan_Outputs_TargetTorque_INPORT" />
    <connection from="bypass.hpMotor_q_OUTPORT"     to="lo_ec_cmdConcate.HeadPan_Outputs_TargetPosition_INPORT" />
    <connection from="bypass.htMotor_q_OUTPORT"     to="lo_ec_cmdConcate.HeadTilt_Outputs_TargetPosition_INPORT" />

<!--
TODO:
    <connection from="bypass.lHand_q_OUTPORT"       to="LeftHand.q_INPORT" />
    <connection from="bypass.lHand_dq_OUTPORT"      to="LeftHand.v_INPORT" />
    <connection from="bypass.lHand_max_i_OUTPORT"   to="LeftHand.t_INPORT" />
    <connection from="bypass.lHand_max_p_OUTPORT"   to="LeftHand.mp_INPORT" />
    <connection from="bypass.lHand_hold_OUTPORT"    to="LeftHand.hold_INPORT" />-->


<!--
TODO:
        this->ports()->addPort("t_OUTPORT", port_t_out_);
        this->ports()->addPort("BHTemp", port_temp_out_);
        this->ports()->addPort("max_measured_pressure_INPORT", port_max_measured_pressure_in_);
        this->ports()->addPort("reset_fingers_INPORT", port_reset_in_);
-->

    <connection from="bypass.rHand_q_OUTPORT"       to="rHand.q_INPORT" />
    <connection from="bypass.rHand_dq_OUTPORT"      to="rHand.v_INPORT" />
    <connection from="bypass.rHand_max_i_OUTPORT"   to="rHand.t_INPORT" />
    <connection from="bypass.rHand_max_p_OUTPORT"   to="rHand.mp_INPORT" />
    <connection from="bypass.rHand_hold_OUTPORT"    to="rHand.hold_INPORT" />
    <connection from="lo_ec_stSplit.CANleft_Box10CANInterfaceIN_RxQueue_OUTPORT" to="rHand.rx_queue_INPORT" />
    <connection from="rHand.tx_OUTPORT"             to="can_queue_tx.tx_INPORT" >
        <conn_policy type="circular_buffer" size="10" />
    </connection>
    <connection from="rHand.q_OUTPORT"              to="hi_stConcate.rHand_q_INPORT" />
    <connection from="rHand.status_OUTPORT"         to="hi_stConcate.rHand_s_INPORT" />

    <connection from="bypass.lHand_q_OUTPORT"       to="lHand.q_INPORT" />
    <connection from="bypass.lHand_dq_OUTPORT"      to="lHand.v_INPORT" />
    <connection from="bypass.lHand_max_i_OUTPORT"   to="lHand.t_INPORT" />
    <connection from="bypass.lHand_max_p_OUTPORT"   to="lHand.mp_INPORT" />
    <connection from="bypass.lHand_hold_OUTPORT"    to="lHand.hold_INPORT" />
    <connection from="lo_ec_stSplit.CANleft_Box10CANInterfaceIN_RxQueue_OUTPORT" to="lHand.rx_queue_INPORT" />
    <connection from="lHand.tx_OUTPORT"             to="can_queue_tx.tx_INPORT" >
        <conn_policy type="circular_buffer" size="10" />
    </connection>
    <connection from="lHand.q_OUTPORT"              to="hi_stConcate.lHand_q_INPORT" />
    <connection from="lHand.status_OUTPORT"         to="hi_stConcate.lHand_s_INPORT" />

    <connection from="can_queue_tx.tx_queue_OUTPORT" to="lo_ec_cmdConcate.CANleft_Box10CANInterfaceOUT_TxQueue_INPORT" />

    <connection from="bypass.sc_OUTPORT"            to="hi_stConcate.sc_INPORT" />
    <connection from="TorsoPanTransmisionInv.MotorCurrent_OUTPORT"  to="bypass.tMotor_i_INPORT" />
    <connection from="hi_cmdSplit.lArm_t_OUTPORT"                   to="bypass.lArm_t_INPORT" />
    <connection from="hi_cmdSplit.rArm_t_OUTPORT"                   to="bypass.rArm_t_INPORT" />
    <connection from="hi_cmdSplit.rHand_q_OUTPORT"                  to="bypass.rHand_q_INPORT" />
    <connection from="hi_cmdSplit.rHand_dq_OUTPORT"                 to="bypass.rHand_dq_INPORT" />
    <connection from="hi_cmdSplit.rHand_max_i_OUTPORT"              to="bypass.rHand_max_i_INPORT" />
    <connection from="hi_cmdSplit.rHand_max_p_OUTPORT"              to="bypass.rHand_max_p_INPORT" />
    <connection from="hi_cmdSplit.rHand_hold_OUTPORT"               to="bypass.rHand_hold_INPORT" />
    <connection from="hi_cmdSplit.lHand_q_OUTPORT"                  to="bypass.lHand_q_INPORT" />
    <connection from="hi_cmdSplit.lHand_dq_OUTPORT"                 to="bypass.lHand_dq_INPORT" />
    <connection from="hi_cmdSplit.lHand_max_i_OUTPORT"              to="bypass.lHand_max_i_INPORT" />
    <connection from="hi_cmdSplit.lHand_max_p_OUTPORT"              to="bypass.lHand_max_p_INPORT" />
    <connection from="hi_cmdSplit.lHand_hold_OUTPORT"               to="bypass.lHand_hold_INPORT" />
    <connection from="HeadPanTransmisionInv.MotorPosition_OUTPORT"  to="bypass.hpMotor_q_INPORT" />
    <connection from="HeadTiltTransmisionInv.MotorPosition_OUTPORT" to="bypass.htMotor_q_INPORT" />


    <connection from="hi_cmdSplit.tMotor_i_OUTPORT" to="TorsoPanTransmisionInv.JointTorque_INPORT" />

    <connection from="packet_counter.packet_counter_OUTPORT"            to="hi_stConcate.test_INPORT" />

    <connection from="lo_rLwr_stSplit.q_OUTPORT"           to="hi_stConcate.rArm_q_INPORT" name="q" />
    <connection from="lo_rLwr_stSplit.dq_OUTPORT"          to="hi_stConcate.rArm_dq_INPORT" name="dq" />
    <connection from="lo_rLwr_stSplit.t_OUTPORT"           to="hi_stConcate.rArm_t_INPORT" name="t" />
    <connection from="lo_rLwr_stSplit.gt_OUTPORT"          to="hi_stConcate.rArm_gt_INPORT" name="gt" />
    <connection from="lo_rLwr_stSplit.w_OUTPORT"           to="hi_stConcate.rArm_w_INPORT" name="wr" />
    <connection from="lo_rLwr_stSplit.mmx_OUTPORT"         to="hi_stConcate.rArm_mmx_INPORT" name="M" />

    <connection from="lo_lLwr_stSplit.q_OUTPORT"           to="hi_stConcate.lArm_q_INPORT" name="q" />
    <connection from="lo_lLwr_stSplit.dq_OUTPORT"          to="hi_stConcate.lArm_dq_INPORT" name="dq" />
    <connection from="lo_lLwr_stSplit.t_OUTPORT"           to="hi_stConcate.lArm_t_INPORT" name="t" />
    <connection from="lo_lLwr_stSplit.gt_OUTPORT"          to="hi_stConcate.lArm_gt_INPORT" name="gt" />
    <connection from="lo_lLwr_stSplit.w_OUTPORT"           to="hi_stConcate.lArm_w_INPORT" name="wr" />
    <connection from="lo_lLwr_stSplit.mmx_OUTPORT"         to="hi_stConcate.lArm_mmx_INPORT" name="M" />

    <connection from="TorsoPanTransmision.JointPosition_OUTPORT"        to="hi_stConcate.tMotor_q_INPORT" name="q" />
    <connection from="TorsoPanTransmision.JointVelocity_OUTPORT"        to="hi_stConcate.tMotor_dq_INPORT" name="dq" />

<!--
TODO:
    <connection from="RightFtSensor.rawWrench_OUTPORT"                  to="hi_stConcate.rFt_rw_INPORT" name="raw_wr" />
    <connection from="RightFtSensor.fastFilteredWrench_OUTPORT"         to="hi_stConcate.rFt_ffw_INPORT" name="fast_wr" />
    <connection from="RightFtSensor.slowFilteredWrench_OUTPORT"         to="hi_stConcate.rFt_sfw_INPORT" name="slow_wr" />
    <connection from="LeftFtSensor.rawWrench_OUTPORT"                   to="hi_stConcate.lFt_rw_INPORT" name="raw_wr" />
    <connection from="LeftFtSensor.fastFilteredWrench_OUTPORT"          to="hi_stConcate.lFt_ffw_INPORT" name="fast_wr" />
    <connection from="LeftFtSensor.slowFilteredWrench_OUTPORT"          to="hi_stConcate.lFt_sfw_INPORT" name="slow_wr" />
-->

    <connection from="lo_rLwr_stSplit.iState_OUTPORT"                         to="hw_state.rArm_fri_state_INPORT" />
    <connection from="lo_rLwr_stSplit.rState_OUTPORT"                       to="hw_state.rArm_robot_state_INPORT" />
    <connection from="lo_lLwr_stSplit.iState_OUTPORT"                         to="hw_state.lArm_fri_state_INPORT" />
    <connection from="lo_lLwr_stSplit.rState_OUTPORT"                       to="hw_state.lArm_robot_state_INPORT" />
<!--
TODO:

    <connection from="LeftHand.q_OUTPORT"               to="hi_stConcate.lHand_q_INPORT" />
    <connection from="LeftHand.status_OUTPORT"          to="hi_stConcate.lHand_s_INPORT" />
-->
<!--
TODO: remove:
    <connection from="ZeroVec8.vector_OUTPORT" to="hi_stConcate.rHand_q_INPORT" name="0" />
    <connection from="Zero.zero_OUTPORT"       to="hi_stConcate.rHand_s_INPORT" />
    <connection from="ZeroVec8.vector_OUTPORT" to="hi_stConcate.lHand_q_INPORT" name="0" />
    <connection from="Zero.zero_OUTPORT"       to="hi_stConcate.lHand_s_INPORT" />
-->

<!--
    <latched_connections>
        <components from="LWRrSim" to="safe" />
        <components from="LWRlSim" to="safe" />
        <components from="TorsoSim" to="safe" />
    </latched_connections>
-->
</subsystem_configuration>
